### PR TITLE
Use allowlisted APIs to fetch information on clusters

### PR DESCRIPTION
**Pull request by smurching**
## What changes are proposed in this pull request?

Use new, allowlisted APIs to fetch information on clusters when DataFrame ACLs are enabled. We fall back to using existing APIs for backwards compatibility with older runtime versions.

## How is this patch tested?

Updated unit tests

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/utils/databricks_utils.py [hunk 1]
```diff
@@ -148,17 +161,25 @@ def get_webapp_url():
...
 
 def get_workspace_info_from_dbutils():
     dbutils = _get_dbutils()
     if dbutils:
-        context = json.loads(
-            dbutils.notebook.entry_point.getDbutils().notebook().getContext().toJson()
-        )
-        workspace_host = context["extraContext"]["api_url"]
-        workspace_id = context["tags"]["orgId"]
+        print("Yo fro msid")
```
Comment (nfx) [inline mlflow/utils/databricks_utils.py:180]: please remove this debug line

**Review by nfx — CHANGES_REQUESTED** [review]

File: mlflow/utils/databricks_utils.py (hunk 1 above)
Comment (smurching) [reply mlflow/utils/databricks_utils.py]: Oops, haha thanks for the catch!

File: mlflow/utils/databricks_utils.py [hunk 2]
```diff
@@ -35,18 +34,29 @@ def _get_java_dbutils():
...
     if tag_opt.isDefined():
         return tag_opt.get()
     else:
         return None
 
 
+def acl_path_of_acl_root():
+    try:
+        return _get_command_context().aclPathOfAclRoot().get()
+    except Exception:  # pylint: disable=broad-except
```
Comment (nfx) [inline mlflow/utils/databricks_utils.py:56]: can you catch exact exception here and remove pylint disable?

File: mlflow/utils/databricks_utils.py [hunk 3]
```diff
@@ -118,7 +128,10 @@ def get_notebook_path():
     path = _get_property_from_spark_context("spark.databricks.notebook.path")
     if path is not None:
         return path
-    return _get_extra_context("notebook_path")
+    try:
+        return _get_command_context().notebookPath().get()
+    except Exception:  # pylint: disable=broad-except
```
Comment (nfx) [inline mlflow/utils/databricks_utils.py:133]: remove `# pylint: disable=broad-except` and catch exact exception

File: mlflow/utils/databricks_utils.py [hunk 4]
```diff
@@ -148,17 +161,24 @@ def get_webapp_url():
...
+    try:
+        return _get_command_context().apiUrl().get()
+    except Exception:  # pylint: disable=broad-except
+        return _get_extra_context("api_url")
+
+
+def get_workspace_id():
+    try:
+        return _get_command_context().workspaceId().get()
+    except Exception:  # pylint: disable=broad-except
```
Comment (nfx) [inline mlflow/utils/databricks_utils.py:173]: remove # pylint: disable=broad-except and catch exact exception

**Review by nfx — CHANGES_REQUESTED** [review]
remove lint ignores and catch exact exceptions

**Comment by smurching** [comment]
@nfx it's a good question whether we can catch a more specific exception type. It turns out Py4j (at least the current release of Py4j) will only ever raise instances of Py4jError ([docs](https://www.py4j.org/advanced_topics.html#py4j-exceptions)). However, since our goal is to support older runtime versions with the current MLflow client (which may have older versions of Py4j or other hard-to-test behavior), it seems the least risky option is to try the new behavior of using allowlisted APIs, and fall back to the existing behavior in case of any exceptions, rather than try to catch a specific exception and find that we've broken behavior on existing runtimes.

One other subtlety worth noting in case we revisit this in the future - `py4j` is intentionally not a dependency of MLflow. In order to catch `Py4jError`, we may be tempted to `from py4j.protocol import Py4jError` at the top level of the `databricks_utils` module, i.e. make it dependent on `py4j`. However since `databricks_utils` is occasionally used in contexts outside of Databricks (e.g. when triggering execution of MLflow projects remotely on Databricks from your laptop), we should instead perform all `py4j` imports inside functions where they're used & ensure these functions are only called on Databricks.

**Comment by nfx** [comment]
@smurching the pain point of catching `Exception` is that it can catch OSError, KeyError, ZeroDivisionError and other dangerous things. if we explicitly don't want to add py4j dependency, then i'd agree with your comment.

**Review by tomasatdatabricks — APPROVED** [review]
lgtm

**Pull request merged**